File: Binary Trees and Binary Search Trees/872. Leaf-Similar Trees/872. Leaf-Similar Trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        leaf1 = []
        leaf2 = []

        def getLeaf(root, arr):
            if not root.left and not root.right:
                arr.append(root.val)
            
            if root.left:
                getLeaf(root.left, arr)
            if root.right:
                getLeaf(root.right, arr)
        
        getLeaf(root1, leaf1) 
        getLeaf(root2, leaf2)

        return leaf1 == leaf2

# Time: O(n + m)
# Space: O(n + m)

File: Binary Trees and Binary Search Trees/872. Leaf-Similar Trees/872. Leaf-Similar Trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root1
 * @param {TreeNode} root2
 * @return {boolean}
 */
var leafSimilar = function (root1, root2) {
  let leaf1 = [];
  let leaf2 = [];

  const getLeaf = (root, arr) => {
    if (!root.left && !root.right) arr.push(root.val);

    if (root.left) getLeaf(root.left, arr);
    if (root.right) getLeaf(root.right, arr);
  };

  getLeaf(root1, leaf1);
  getLeaf(root2, leaf2);

  return (
    leaf1.length === leaf2.length && leaf1.every((val, i) => val === leaf2[i])
  );
};

// Time: O(n + m)
// Space: O(n + m)
